### DISCUSSION TITLE
Adding option to allow "machine" interfacing

### DISCUSSION THREAD

**Discussion by lombare** [Exiv2/exiv2#3058] [Q&A] [answered]
Hi, here ! 

I'm currently developing a little app where I need to extract all images metadata etc ...

I was wondering if there was a way (and if not what is your thought on adding this feature, I'm okkay with working on it), to "print" theses datas in "machine readable" format ?

Also (because in my case it'll be simpler), what do you think of adding a way for exiv to read the picture on Stdin (like `grep ... - ` does), it would avoid in this case, creating a temporary file

**Comment by clanmills** [comment, marked as answer]
Exiv2 can do all of those things. It can print metadata as "value" _(raw data)_ or "translated" _(human readable)_.

```
$ exiv2 -pv -g datetime/i -g orientation/i https://clanmills.com/Stonehenge.jpg
0x0112 Image        Orientation                 Short       1  1
0x0132 Image        DateTime                    Ascii      20  2015:07:16 20:25:28
0x9003 Photo        DateTimeOriginal            Ascii      20  2015:07:16 15:38:54
0x9004 Photo        DateTimeDigitized           Ascii      20  2015:07:16 15:38:54
$ exiv2 -pt -g datetime/i -g orientation/i https://clanmills.com/Stonehenge.jpg
Exif.Image.Orientation                       Short       1  top, left
Exif.Image.DateTime                          Ascii      20  2015:07:16 20:25:28
Exif.Photo.DateTimeOriginal                  Ascii      20  2015:07:16 15:38:54
Exif.Photo.DateTimeDigitized                 Ascii      20  2015:07:16 15:38:54
$ 
```
It can read images from stdin (and write to stdout)
```
$ curl https://clanmills.com/Stonehenge.jpg | exiv2 -g datetime/i -g orientation/i -
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 6599k  100 6599k    0     0   637k      0  0:00:10  0:00:10 --:--:-- 1302k
Exif.Image.Orientation                       Short       1  top, left
Exif.Image.DateTime                          Ascii      20  2015:07:16 20:25:28
Exif.Photo.DateTimeOriginal                  Ascii      20  2015:07:16 15:38:54
Exif.Photo.DateTimeDigitized                 Ascii      20  2015:07:16 15:38:54
$ 
```
Exiv2 is optimized to read metadata from remote locations using HTTP ByteRange to avoid copying the complete file.

The code in the exiv2 application that prints the output above is in src/actions.cpp
```
        if (Params::instance().printItems_ & Params::prValue && md.size() > 0) {
            if (!first)
                std::cout << "  ";
            first = false;
            std::ostringstream os;
            // #1114 - show negative values for SByte
            if (md.typeId() == Exiv2::signedByte) {
                for ( long c = 0 ; c < md.value().count() ; c++ ) {
                    long value = md.value().toLong(c);
                    os << (c?" ":"") << std::dec << (value < 128 ? value : value - 256);
                }
            } else {
                os << std::dec << md.value();
            }
            binaryOutput(os);
        }
        if (Params::instance().printItems_ & Params::prTrans) {
            if (!first)
                std::cout << "  ";
            first = false;
            std::ostringstream os;
            os << std::dec << md.print(&pImage->exifData());
            binaryOutput(os) ;
        }
```
The API `os << md.print();` implements "translated", and the "raw" is simply `os << md.value()`.  If you want to value for purposes other that printing, you should use `md.value()`.

If you're happy with my answer, please mark this question as answered.  As I have retired and I don't subscribe to github, please email if you wish to discuss this further.  robinwmills@gmail.com

**Reply by lombare** [reply]
Sorry for my late reply I took time off for vacations.

Thanks for your reply, looks like it answer my question, I will try this tonight !
I couldn't find information in the help about stdin (this is exactly what I needed) so I wrongly assumed it wasn't supported

Thanks :)

**Reply by clanmills** [reply]
@lombare.  @postscript-dev has done wonderful work on the exiv2 documentation.  

The exiv2 command-line utility is a _**swiss army knife**_ with many blades to remove your fingers.  It is the backbone of the exiv2 test harness.  However, with so many options, it's easy to miss features in the man page.
